paddle/fluid/operators/sgd_op.cc

```diff
@@ -43,20 +43,32 @@ class SGDOp : public framework::OperatorWithKernel {
  protected:
   framework::OpKernelType GetExpectedKernelType(
       const framework::ExecutionContext& ctx) const override {
-    return framework::OpKernelType(
-        framework::ToDataType(ctx.Input<framework::LoDTensor>("Param")->type()),
-        ctx.GetPlace());
+    auto* table_var = ctx.InputVar("Param");
+    if (table_var->IsType<framework::LoDTensor>()) {
+      return framework::OpKernelType(
+          framework::ToDataType(table_var->Get<framework::LoDTensor>().type()),
+          ctx.device_context());
+    } else if (table_var->IsType<framework::SelectedRows>()) {
+      return framework::OpKernelType(
+          framework::ToDataType(
+              table_var->Get<framework::SelectedRows>().value().type()),
+          ctx.device_context());
+    } else {
+      PADDLE_THROW("Param should be LoDTensor or SelectedRows");
+    }
```

Comment:
line46-58 is similar to line21-35 of lookup_tabel_op, so I think we should add a new API, like `GetPODTypeFromVar`, do you think so?

Reply:
done